java/org/apache/tomcat/util/IntrospectionUtils.java

```diff
@@ -332,7 +341,15 @@ public static String replaceProperties(String value,
         }
         if (prev < value.length())
             sb.append(value.substring(prev));
-        return sb.toString();
+        String newval = sb.toString();
+        if (newval.indexOf('$') < 0) {
+            return newval;
+        }
+        if (newval.equals(value))
+            return value;
+        if (log.isDebugEnabled())
+            log.debug("IntrospectionUtils.replaceProperties iter on: " + newval);
+        return replaceProperties(newval, staticProp, dynamicProp, classLoader, iterationCount+1);
```

Comment:
Does this need to be recursive? Can we not simply have a loop of replacements?

Reply:
A loop would work to but that is more changes.

Reply:
I would stay with the recursion here, as it is doesn't seem to be time critical to replace those placeholders and I find it easier to read with the recursion.